force-app/main/default/classes/DMLOperations.cls

```diff
@@ -145,76 +179,150 @@ public with sharing class DMLOperations {
-
-		// Link the Contact to its Account using the Account Id
-
-		// Upsert the Contacts
-	}
+        for ( Contact cont : contactsToUpsert) {
+            // Extract the Account Name from Contact's LastName
+            accName = cont.Lastname;
+            contNames.add(accName);
+        }
+        // Get the account based on the name or create one if it doesn't exist
+        List <Account> nameList = [SELECT Name, Id FROM Account WHERE Name = :contNames];
+
+        Map<String, Id> actMap = new Map<String, Id>();
+        for (Account acc : nameList) {
+            actMap.put(acc.Name, acc.Id);
+            // Link the Contact to its Account using the Account Id
+        } 
+        
+        for (Contact cont1 : contactsToUpsert){
+            cont1.AccountId = actMap.get(cont1.LastName);
+            if ( cont1.AccountId == null ){
+                Account newAcc = upsertAccount(cont1.LastName);
+                cont1.AccountId = newAcc.Id;
+
+            }
+        } // Upsert the Contacts 
+        upsert contactsToUpsert;           
+	} 
 
     /**
      * Question 11
      * Insert a list of Leads and then deletes them.
      * @param leadNames The list of Lead names to be inserted.
      */
-    public static void insertAndDeleteLeads(List<String> leadNames) {
+     public static void insertAndDeleteLeads(List<String> leadNames) {
+        List<Lead> lds = new List<Lead>();
+        String company = 'Test Company';
         // Loop through each Lead name
+        for ( String leadName :leadNames ) {
+            List<String> nameParts = leadName.split(' ');
+            Lead l = new Lead();
+            l.FirstName = nameParts[0];
+            l.LastName = nameParts[1];
+            l.Company = company;
+            lds.add(l);
+        }
 
         // Insert the Leads
+        insert lds;
 
         // Delete the Leads
+        List<Lead> deleteList = [SELECT Name FROM Lead WHERE Company = :company];
+        Database.delete(deleteList);
+
     }
 
 
     /**
      * Question 12
-     * Dynamically create the number of cases and relate it to the accountId, then deletes them.
+     * Dynamically creates the number of cases and relate it to the accountId, then deletes them.
      * @param accountId The account for which the cases will be created and then deleted.
      * @param numOfCases The number of cases to create.
      */
     public static void createAndDeleteCases(String accountId, Integer numOfCases) {
         // Insert the cases
-
+        List<Case> cases = new List<Case>();
+        Case cs = new Case();
+        for ( Integer i = 0; i < numOfCases; i++ ) {
+            cs.Subject = 'test case';
+            cs.Status = 'New';
+            cs.Origin = 'Phone';
+            cases.add(cs);
+            System.debug(cases);
+        } 
         // Delete the cases
+        List<Case> deleteCase = [SELECT Id FROM Case WHERE Subject = 'test case'];
+        Database.delete(deleteCase);
     }
```

Comment:
This test passes because the cases were not actually inserted. Do a dml and insert the cases before delete

Reply:
I think I must have gone back to work on another problem and just forgot to go back and complete this one because there was no failures. I also had not related the cases to the Accounts.